File: learnLeetCode_remote/Day15/Part4/largest_rectangle.py
from typing import List

# LeetCode 84 — Largest Rectangle in Histogram
# Monotonic stack, append sentinel 0 to flush.

def largestRectangleArea(heights: List[int]) -> int:
    stack = []  # indices of increasing heights
    res = 0
    h = heights + [0]
    for i, ht in enumerate(h):
        while stack and h[stack[-1]] > ht:
            H = h[stack.pop()]
            L = stack[-1] if stack else -1
            width = i - L - 1
            res = max(res, H * width)
        stack.append(i)
    return res

if __name__ == "__main__":
    print(largestRectangleArea([2,1,5,6,2,3]))  # 10

File: learnLeetCode/Day15/Part4/largest_rectangle.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode 84 — Largest Rectangle in Histogram

int largestRectangleArea(vector<int>& heights){
    vector<int> h = heights; h.push_back(0);
    vector<int> st; long long ans=0;
    for (int i=0;i<(int)h.size();++i){
        while (!st.empty() && h[st.back()] > h[i]){
            int idx = st.back(); st.pop_back();
            int L = st.empty() ? -1 : st.back();
            long long width = i - L - 1;
            ans = max(ans, 1LL*h[idx]*width);
        }
        st.push_back(i);
    }
    return (int)ans;
}

int main(){
    vector<int> a = {2,1,5,6,2,3};
    cout << largestRectangleArea(a) << "\n"; // 10
    return 0;
}
